src/ratelimiter.rs

```diff
@@ -0,0 +1,45 @@
+use std::thread::sleep;
+use std::time::{Duration, Instant};
+
+#[derive(Debug)]
+pub struct RateLimiter {
+    available_tokens: f32,
+    tokens_per_second: f32,
+    last_update: Instant,
+}
+
+impl RateLimiter {
+    pub(crate) fn new(chunk_size_bytes: u64, mbps_target: Option<u64>) -> Self {
+        let bps_target = mbps_target.map(|x| x * 1_000_000);
+        let tokens_per_second = match bps_target {
+            None => 0.0,
+            Some(bps) => bps as f32 / chunk_size_bytes as f32,
+        };
+        RateLimiter {
+            available_tokens: 0.0,
```

Comment:
Typically the bucket starts out full, not empty. If it starts out empty, the very first send we try to do will block unnecessarily.

Reply:
Initially I started with a full bucket, but doing that you can send 2x the bucket size over the first transfer window.

I'm not sure that it's an actual problem, but I thought that if we are limiting the max bandwidth, we shouldn't exceed the threshold (instead of starting at 2x and gradually trending downwards to the threshold)

What do you think?

Reply:
> but doing that you can send 2x the bucket size over the first transfer window.

I don’t think so, but it’s a bit subtle. I think pictures make it clearer. This is what an ideal transfer would look like:

![send-ceil](https://github.com/user-attachments/assets/3fbab400-a8a4-4bdd-9eec-5892cddb23d5)

Where on the y-axis is cumulative bytes sent (in arbitrary units, where 1 means 1 bucket, not necessarily 1 byte), and on the x-axis is time. It’s not so clear in the plot but the intervals should be closed on the left and open on the right, i.e. `f(0) = 1`, `f(1 – ε) = 1`, `f(1) = 2`, etc.

The instantaneous transfer rate is the slope of this function, but that’s not a useful thing to look at because the derivative is zero everywhere and it’s not differentiable at the discontinuities. (Which just means that if you measure over a small enough time window, then when there is no send in that window the rate is 0, and when there is a send in that window, you can make the rate arbitrarily high by making the window arbitrarily small.)

So what we want to look at instead is the average transfer rate over some time window, `f(t1) – f(t0) / (t1 - t0)`.

In the limit where `t1 - t0 → ∞`, this rate goes to 1 (for the function I plotted, but again the y-axis units are arbitrary). But rate limits in networks don’t work over infinite time windows so this is _also_ not so useful to look at.

So then what time window do we take? Well, if we take a time window of `t0=0, t1=1`, then we get `f(1) – f(0) / (1 – 0) = (2 – 1) / 1 = 1`. So we only count one rising edge inside the window. But if we take the limit `ε → 0` for the window `t0=–ε, t1=1`, then we get `f(–ε) = 0` so the rate is 2. So for a window of size 1, the rate is 1, but if we make the window the tiniest amount larger so it can fit two rising edges, the rate is suddenly 2.

If we plot the average transfer rate over a window `t0=0, t1=x`, it looks like this:

![transfer-rate](https://github.com/user-attachments/assets/6530bcc1-b6ba-45aa-b442-d8de250daf7b)

> starting at 2x and gradually trending downwards to the threshold

So sure, it looks like the rate is high and then drops down to the target rate, but that’s an artifact caused by how we define the rate, the pattern of how we transfer does not change over time! If we only evaluate this rate function at integer times then we find a rate of 1 everywhere, but if we evaluate it at an ε past integers then we find a rate of `1 + 1/x` and we have this “gradually trending downwards towards the threshold”.

Now for my next magic trick, let’s look at the average transfer rate over a window `ε → 0 t0=ε, t1=x`, so now this initial rising edge is not part of the window. Then we get this:

![rate-window-epsilon](https://github.com/user-attachments/assets/1aabf55a-9e70-44bb-98db-939d032dceb5)

Again, when we evaluate it at integer times then we find a rate of 1 everywhere (`x=0` does not count, the function is not defined for x < ε). But if we evaluate it at an ε before integers then we find a rate of `1 - 1/x`. Now it starts below the threshold and it’s gradually trending up!

So when you say

> but doing that you can send 2x the bucket size over the first transfer window.

I think we agree on what the send pattern should be, it’s just that we are talking about different windows to measure the rate over. If the bucket starts out empty, then initially we have to wait and cumulative bytes sent looks like this:

![cumulative-sent](https://github.com/user-attachments/assets/7a0010a3-6f8d-4ebd-90ba-8fe3e76af005)

Which is exactly the same graph as before, just shifted in time, so the rate should also be the same, just shifted in time.

Reply:
Plotting the instant transfer rate (rate over a very small window) along with the average helped me understand that there are only two rates:
- While bucket is full, we transfer at the max bandwidth of the underlying connection
- While the bucket is "empty" (filling at its standard rate) we transfer at the max bandwidth that the ratelimiter allows

When the max bandwidth is over the target bandwidth (3.5mbps, 3mbps respectively here):
![image](https://github.com/user-attachments/assets/20d1cf70-c1f2-4a08-85bd-5c617b8ad6de)




Starting with an empty bucket for comparison
![image](https://github.com/user-attachments/assets/9378083d-3d04-46bf-a5f4-039051e4b173)

shows the same behavior, only that we are indeed blocked until the bucket refills enough for the first transfer

Reply:
I think I assumed that the capacity of the bucket is on the same order of magnitude as what we send in one send call. If it’s smaller then we wouldn’t send anything, if it’s much larger then yes initially we are not limiting and sending as fast as the network allows. That’s a good way to determine the capacity of the bucket then, it should be about one chunk.